### DISCUSSION TITLE
sense2vec factory issue

### DISCUSSION THREAD

**Discussion by sdspieg** [explosion/spaCy#10720] [Help: Other Questions] [answered]
Following [this blog item](https://explosion.ai/blog/sense2vec-reloaded) in Google Colab (with high ram) to try out sense2vec, I run into this error. Any suggestions on how to fix this? 

Here is the code cell:

```python
nlp = spacy.load("en_core_web_lg")
s2v = Sense2VecComponent(nlp.vocab).from_disk("/content/drive/MyDrive/RuBase/Corpora/sense2vec/s2v_reddit_2019_lg")
nlp.add_pipe("s2v")`
```

And here's the error message:

```python-traceback
ValueError                                Traceback (most recent call last)
[<ipython-input-7-7af9fb234ba5>](https://localhost:8080/#) in <module>()
      2 nlp = spacy.load("en_core_web_lg")
      3 s2v = Sense2VecComponent(nlp.vocab).from_disk("/content/drive/MyDrive/RuBase/Corpora/sense2vec/s2v_reddit_2019_lg")
----> 4 nlp.add_pipe("s2v")

1 frames
[/usr/local/lib/python3.7/dist-packages/spacy/language.py](https://localhost:8080/#) in create_pipe(self, factory_name, name, config, raw_config, validate)
    653                 lang_code=self.lang,
    654             )
--> 655             raise ValueError(err)
    656         pipe_meta = self.get_factory_meta(factory_name)
    657         # This is unideal, but the alternative would mean you always need to

ValueError: [E002] Can't find factory for 's2v' for language English (en). This usually happens when spaCy calls `nlp.create_pipe` with a custom component name that's not registered on the current language class. If you're using a Transformer, make sure to install 'spacy-transformers'. If you're using a custom component, make sure you've added the decorator `@Language.component` (for function components) or `@Language.factory` (for class components).

Available factories: attribute_ruler, tok2vec, merge_noun_chunks, merge_entities, merge_subtokens, token_splitter, doc_cleaner, parser, beam_parser, entity_linker, ner, beam_ner, entity_ruler, lemmatizer, tagger, morphologizer, senter, sentencizer, textcat, spancat, textcat_multilabel, sense2vec, en.lemmatizer
```

**Comment by koaning** [comment, marked as answer]
It may be that you're running into a version mismatch. What version of spaCy and sense2vec are you using? 

The [sense2vec](https://github.com/explosion/sense2vec) docs mention the usage for spaCy v3, if you're interested in using an older version, you should check [this branch of the project](https://github.com/explosion/sense2vec/tree/v1.x).

**Comment by sdspieg** [comment]
Thanks for that quick response. I am using that page and am trying to replicate it in Google colab. So I did indeed have spaCy 2.2.4 . I uninstalled it and `python -m spacy info` now tells me I am running 3.2.4 
But I still get the same error
```ValueError                                Traceback (most recent call last)
[<ipython-input-4-7af9fb234ba5>](https://localhost:8080/#) in <module>()
      2 nlp = spacy.load("en_core_web_lg")
      3 s2v = Sense2VecComponent(nlp.vocab).from_disk("/content/drive/MyDrive/RuBase/Corpora/sense2vec/s2v_reddit_2019_lg")
----> 4 nlp.add_pipe("s2v")

1 frames
[/usr/local/lib/python3.7/dist-packages/spacy/language.py](https://localhost:8080/#) in create_pipe(self, factory_name, name, config, raw_config, validate)
    653                 lang_code=self.lang,
    654             )
--> 655             raise ValueError(err)
    656         pipe_meta = self.get_factory_meta(factory_name)
    657         # This is unideal, but the alternative would mean you always need to

ValueError: [E002] Can't find factory for 's2v' for language English (en). This usually happens when spaCy calls `nlp.create_pipe` with a custom component name that's not registered on the current language class. If you're using a Transformer, make sure to install 'spacy-transformers'. If you're using a custom component, make sure you've added the decorator `@Language.component` (for function components) or `@Language.factory` (for class components).

Available factories: attribute_ruler, tok2vec, merge_noun_chunks, merge_entities, merge_subtokens, token_splitter, doc_cleaner, parser, beam_parser, entity_linker, ner, beam_ner, entity_ruler, lemmatizer, tagger, morphologizer, senter, sentencizer, textcat, spancat, textcat_multilabel, sense2vec, en.lemmatizer```

**Reply by koaning** [reply]
Did you remember to restart the notebook? I'm running the code for the 2015 embeddings, and these run just fine. 

## Step 1

Install the recent versions of the packages. 

```
! python -m pip install --upgrade spacy sense2vec
```

## Step 2

Restart the notebook. This is very important! 

## Step 3 

Download the spaCy model and sense2vec model. 

```
! python -m spacy download en_core_web_sm
! wget https://github.com/explosion/sense2vec/releases/download/v1.0.0/s2v_reddit_2015_md.tar.gz
```

Once the s2v model is in, make sure to untar it. 

```
! tar -xf s2v_reddit_2015_md.tar.gz
```

## Step 4: Load Code 

This runs fine now: 

```python
from sense2vec import Sense2Vec
s2v = Sense2Vec().from_disk("s2v_old")
```

Same here: 

```python
import spacy
from sense2vec import Sense2VecComponent

nlp = spacy.load("en_core_web_sm")
s2v = nlp.add_pipe("sense2vec")
s2v.from_disk("s2v_old")
```

**Reply by sdspieg** [reply]
Thanks for your patience. And these instructions are already more useful to quasi-noobs like myself, as they start from the very beginning (e.g. they may not have spacy or even python and pip installed; they may not know what has to be run in ubuntu and what in python, etc.) and then go step by step. I really think it'd be great if the spaCy docu would actually contain such noob-proof Google colabs or Jupyter notebooks. But so I'm trying this out - will report back

**Reply by sdspieg** [reply]
Reporting back. I tried doing this with the en_core_web_lg NLP-model and with the s2v_reddit_2019_lg transformers. That does seem to work now, but only until the code cell that I have that ends with 
`s2v.from_disk("s2v_reddit_2019_lg")` 
The output says <sense2vec.component.Sense2VecComponent at 0x7f0cc3ca1a90>, which I guess means it went through.
But when I then do 
`doc = nlp("A sentence about natural language processing.")
assert doc[3:6].text == "natural language processing"
freq = doc[3:6]._.s2v_freq
vector = doc[3:6]._.s2v_vec
most_similar = doc[3:6]._.s2v_most_similar(3)`, I get
---------------------------------------------------------------------------

ValueError                                Traceback (most recent call last)
[<ipython-input-12-da4eb69238c4>](https://localhost:8080/#) in <module>()
----> 1 doc = nlp("A sentence about natural language processing.")
      2 assert doc[3:6].text == "natural language processing"
      3 freq = doc[3:6]._.s2v_freq
      4 vector = doc[3:6]._.s2v_vec
      5 most_similar = doc[3:6]._.s2v_most_similar(3)

5 frames
[/usr/local/lib/python3.7/dist-packages/spacy/language.py](https://localhost:8080/#) in __call__(self, text, disable, component_cfg)
   1020                 raise ValueError(Errors.E109.format(name=name)) from e
   1021             except Exception as e:
-> 1022                 error_handler(name, proc, [doc], e)
   1023             if doc is None:
   1024                 raise ValueError(Errors.E005.format(name=name))

[/usr/local/lib/python3.7/dist-packages/spacy/util.py](https://localhost:8080/#) in raise_error(proc_name, proc, docs, e)
   1615 
   1616 def raise_error(proc_name, proc, docs, e):
-> 1617     raise e
   1618 
   1619 

[/usr/local/lib/python3.7/dist-packages/spacy/language.py](https://localhost:8080/#) in __call__(self, text, disable, component_cfg)
   1015                 error_handler = proc.get_error_handler()
   1016             try:
-> 1017                 doc = proc(doc, **component_cfg.get(name, {}))  # type: ignore[call-arg]
   1018             except KeyError as e:
   1019                 # This typically happens if a component is not initialized

[/usr/local/lib/python3.7/dist-packages/sense2vec/component.py](https://localhost:8080/#) in __call__(self, doc)
    102         """
    103         if self.first_run:
--> 104             self.init_component()
    105             self.first_run = False
    106         # Store reference to s2v object on Doc to make sure it's right

[/usr/local/lib/python3.7/dist-packages/sense2vec/component.py](https://localhost:8080/#) in init_component(self)
    118         not added.
    119         """
--> 120         Doc.set_extension("_s2v", default=None)
    121         Doc.set_extension("s2v_phrases", getter=self.get_phrases)
    122         for obj in [Token, Span]:

/usr/local/lib/python3.7/dist-packages/spacy/tokens/doc.pyx in spacy.tokens.doc.Doc.set_extension()

ValueError: [E090] Extension '_s2v' already exists on Doc. To overwrite the existing extension, set `force=True` on `Doc.set_extension`.

**Reply by sdspieg** [reply]
UPD - I reran the entire [Colab notebook](https://colab.research.google.com/drive/1jHIrTzILqfuwEu79Zd-l8p5QfYB3JLpX?usp=sharing) I made based on your recommendations, but with the web_lg model and the 2019 reddit vectors. It did run through without any errors this time, but the last code cell did not generate any output 
![image](https://user-images.githubusercontent.com/691472/165863516-11978da3-eba1-4305-865d-d2c9d68d2a79.png)

**Reply by koaning** [reply]
That's normal because the output of the code is stored in variables.

If you're rather new at Python, my best recommendation is to spend some time mastering it. It takes a while, but it's much easier to master the spaCy library if you understand how Python works. If you're looking for resources; I hear good things from [freecodecamp](https://www.freecodecamp.org/), and I also host many personal tutorials on [calmcode.io](https://calmcode.io).

Also, if you're interested in running sense2vec, we also have a [limited online demo](https://explosion.ai/demos/sense2vec) that doesn't require you to write Python.

**Reply by sdspieg** [reply]
Thank you. But in the meanwhile, could you still please tell me (/us) how I (/we) can input text strings and obtain semantically similar text spans? As in [the blog entry](https://github.com/explosion/sense2vec)
```doc = nlp("A sentence about natural language processing.")
assert doc[3:6].text == "natural language processing"
freq = doc[3:6]._.s2v_freq
vector = doc[3:6]._.s2v_vec
most_similar = doc[3:6]._.s2v_most_similar(3)
# [(('machine learning', 'NOUN'), 0.8986967),
#  (('computer vision', 'NOUN'), 0.8636297),
#  (('deep learning', 'NOUN'), 0.8573361)]```

**Reply by koaning** [reply]
One small thing, could you apply markdown formatting on codeblocks? That makes it *much* easier to read. You can apply markdown formatting by editing your comment cell and by wrapping your code with three ticks "```". 

Here's an example; 

````md
This is normal text. 

```python
# This cell contains python code

doc = nlp("A sentence about natural language processing.")
```
````

**Reply by koaning** [reply]
> But in the meanwhile, could you still please tell me (/us) how I (/we) can input text strings and obtain semantically similar text spans?

I believe your code already does that if you open up a new cell with just this variable: 

```
most_similar
```

... then when you run the cell it should print the variable contents, which in this case I expect to look something like: 

```
[(('machine learning', 'NOUN'), 0.8986967),
  (('computer vision', 'NOUN'), 0.8636297),
  (('deep learning', 'NOUN'), 0.8573361)]
```

**Reply by sdspieg** [reply]
> One small thing, could you apply markdown formatting on codeblocks? That makes it _much_ easier to read. You can apply markdown formatting by editing your comment cell and by wrapping your code with three ticks "```".

I now did. But I was just following what I saw in the editing ribbon, and that just suggests one tick. Maybe that could be fixed?

**Reply by sdspieg** [reply]
> I believe your code already does that if you open up a new cell with just this variable:
> 
> ```
> most_similar
> ```

Thanks Vincent. That did the trick... 
![image](https://user-images.githubusercontent.com/691472/166072464-9fa95148-bcf2-4a9a-b2b6-d9a524e717be.png)
Maybe the team could consider adding that one line to the docu? 
So now I'll have to start digging into the documentation on how we can start training our own vectors and how to use this to generate noun phrases and bootstrapping NER annotations in Prodigy. 
I so hope that people will also start making and sharing FULL Google colabs or Jupyter notebooks for all of these things, where noobs like me can then FIRST successfully run the entire demo on their own systems to see what happens and THEN plug in their own data/corpora to obtain similar results in the fields they are working in. That would IMO REALLY mean 'democratization'...
